File: go/logger.go
package commonlog

import (
	"log"

	"github.com/alvianhanif/commonlog/go/providers"
	"github.com/alvianhanif/commonlog/go/types"
)

// ====================
// Main Logger
// ====================

// createProvider creates a provider instance by name
func createProvider(providerName string) types.Provider {
	switch providerName {
	case "slack":
		return &providers.SlackProvider{}
	case "lark":
		return &providers.LarkProvider{}
	default:
		return &providers.SlackProvider{}
	}
}

// Logger is the main struct
type Logger struct {
	config   types.Config
	provider types.Provider
}

// NewLogger creates a new Logger with the appropriate provider
func NewLogger(cfg types.Config) *Logger {
	providerName := createProvider(cfg.Provider)
	return &Logger{config: cfg, provider: providerName}
}

// resolveChannel resolves the channel for the given alert level
func (l *Logger) resolveChannel(level int) string {
	if l.config.ChannelResolver != nil {
		return l.config.ChannelResolver.ResolveChannel(level)
	}
	return l.config.Channel
}

// Send sends a message with alert level, optional attachment, and optional trace log
func (l *Logger) Send(level int, message string, attachment *types.Attachment, trace string) error {
	return l.SendToChannel(level, message, attachment, trace, "")
}

// SendToChannel sends a message to a specific channel, overriding the default/channel resolver
func (l *Logger) SendToChannel(level int, message string, attachment *types.Attachment, trace string, channel string) error {
	if level == types.INFO {
		log.Printf("[INFO] %s", message)
		return nil
	}

	resolvedChannel := channel
	if resolvedChannel == "" {
		resolvedChannel = l.resolveChannel(level)
	}

	sendConfig := l.config
	sendConfig.Channel = resolvedChannel

	if trace != "" {
		traceAttachment := &types.Attachment{
			FileName: "trace.log",
			Content:  trace,
		}
		if attachment != nil {
			if attachment.Content != "" {
				attachment.Content += "\n\n--- Trace Log ---\n" + trace
			} else {
				attachment.Content = trace
				attachment.FileName = "trace.log"
			}
		} else {
			attachment = traceAttachment
		}
	}

	return l.provider.SendToChannel(level, message, attachment, sendConfig, resolvedChannel)
}

// CustomSend sends a message with a custom provider, allowing override of the default provider
func (l *Logger) CustomSend(provider string, level int, message string, attachment *types.Attachment, trace string, channel string) error {
	customProvider := createProvider(provider)
	if customProvider == nil {
		log.Printf("[ERROR] Unknown provider: %s, defaulting to slack", provider)
		customProvider = createProvider("slack")
	}

	if level == types.INFO {
		log.Printf("[INFO] %s", message)
		return nil
	}

	resolvedChannel := channel
	if resolvedChannel == "" {
		resolvedChannel = l.resolveChannel(level)
	}

	sendConfig := l.config
	sendConfig.Channel = resolvedChannel

	if trace != "" {
		traceAttachment := &types.Attachment{
			FileName: "trace.log",
			Content:  trace,
		}
		if attachment != nil {
			if attachment.Content != "" {
				attachment.Content += "\n\n--- Trace Log ---\n" + trace
			} else {
				attachment.Content = trace
				attachment.FileName = "trace.log"
			}
		} else {
			attachment = traceAttachment
		}
	}

	return customProvider.SendToChannel(level, message, attachment, sendConfig, resolvedChannel)
}

File: python/logger.py
"""
Main logger for commonlog
"""
import logging
from .providers import SlackProvider, LarkProvider
from .log_types import AlertLevel, Attachment

# ====================
# Configuration and Logger
# ====================

class commonlog:
    def send_to_channel(self, level, message, attachment=None, trace="", channel=None):
        if level == AlertLevel.INFO:
            logging.info(message)
            return
        try:
            # Use provided channel or fallback to resolved channel
            target_channel = channel if channel else self._resolve_channel(level)
            original_channel = self.config.channel
            self.config.channel = target_channel
            if trace:
                if attachment is None:
                    attachment = Attachment(content=trace, file_name="trace.log")
                else:
                    if attachment.content:
                        attachment.content += "\n\n--- Trace Log ---\n" + trace
                    else:
                        attachment.content = trace
                        attachment.file_name = "trace.log"
            self.provider.send(level, message, attachment, self.config)
            self.config.channel = original_channel
        except Exception as e:
            logging.error(f"Failed to send alert: {e}")
            raise

    def custom_send(self, provider, level, message, attachment=None, trace="", channel=None):
        if provider == "slack":
            custom_provider = SlackProvider()
        elif provider == "lark":
            custom_provider = LarkProvider()
        else:
            logging.warning(f"Unknown provider: {provider}, defaulting to Slack")
            custom_provider = SlackProvider()

        if level == AlertLevel.INFO:
            logging.info(message)
            return
        try:
            # Use provided channel or fallback to resolved channel
            target_channel = channel if channel else self._resolve_channel(level)
            original_channel = self.config.channel
            self.config.channel = target_channel
            if trace:
                if attachment is None:
                    attachment = Attachment(content=trace, file_name="trace.log")
                else:
                    if attachment.content:
                        attachment.content += "\n\n--- Trace Log ---\n" + trace
                    else:
                        attachment.content = trace
                        attachment.file_name = "trace.log"
            custom_provider.send(level, message, attachment, self.config)
            self.config.channel = original_channel
        except Exception as e:
            logging.error(f"Failed to send alert: {e}")
            raise

    def __init__(self, config):
        self.config = config
        if config.provider == "slack":
            self.provider = SlackProvider()
        elif config.provider == "lark":
            self.provider = LarkProvider()
        else:
            logging.warning(f"Unknown provider: {config.provider}, defaulting to Slack")
            self.provider = SlackProvider()

    def _resolve_channel(self, level):
        if self.config.channel_resolver:
            return self.config.channel_resolver.resolve_channel(level)
        return self.config.channel

    def send(self, level, message, attachment=None, trace=""):
        if level == AlertLevel.INFO:
            logging.info(message)
            return
        try:
            # Resolve the channel for this alert level
            resolved_channel = self._resolve_channel(level)
            
            # Temporarily modify config with resolved channel
            original_channel = self.config.channel
            self.config.channel = resolved_channel
            
            # If trace is provided, create an attachment
            if trace:
                if attachment is None:
                    attachment = Attachment(content=trace, file_name="trace.log")
                else:
                    # If there's already an attachment, combine the trace content
                    if attachment.content:
                        attachment.content += "\n\n--- Trace Log ---\n" + trace
                    else:
                        attachment.content = trace
                        attachment.file_name = "trace.log"
            self.provider.send(level, message, attachment, self.config)
            
            # Restore original channel
            self.config.channel = original_channel
        except Exception as e:
            logging.error(f"Failed to send alert: {e}")
            raise